### PR TITLE
journal: adapt for new improved LZ4_decompress_safe_partial()

### LINKED ISSUE #10259: journal: LZ4_decompress_safe_partial() is fixed in lz4-1.8.3
**systemd version the issue has been seen with**
> 239 / 9380d34c2a661f071cbdabe24f39d3b63ab50f8f

**Used distribution**
 > Arch Linux

**Expected behaviour you didn't see**
 > meson test -C build to pass

**Unexpected behaviour you saw**
 ```
256/340 test-compress                           FAIL     0.36 s (killed by signal 6 SIGABRT)

--- command ---
SYSTEMD_LANGUAGE_FALLBACK_MAP='/build/systemd/src/systemd/src/locale/language-fallback-map' PATH='/build/systemd/src/build:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl' SYSTEMD_KBD_MODEL_MAP='/build/systemd/src/systemd/src/locale/kbd-model-map' /build/systemd/src/build/test-compress
--- stderr ---
Failed to read $container of PID 1, ignoring: Permission denied
Found container virtualization none.
/* testing XZ text blob compression/decompression */
/* testing XZ random blob compression/decompression */
compression failed: No buffer space available
/* testing decompress_startswith with XZ on text text */
/* testing decompress_startswith with XZ on random text */
compression failed: No buffer space available
/* testing decompress_startswith with XZ on HUGE=xxxxxxxxxxxxxxx text */
compression failed: No buffer space available
/* testing XZ compression */
/* create source from /build/systemd/src/build/test-compress */
/* test compression */
XZ compression finished (22288 -> 5868 bytes, 26.3%)
/* test decompression */
XZ decompression finished (5868 -> 22288 bytes, 379.8%)
/* test faulty decompression */
Decompression failed: code 7
/* testing LZ4 text blob compression/decompression */
/* testing LZ4 random blob compression/decompression */
compression failed: No buffer space available
/* testing decompress_startswith with LZ4 on text text */
/* testing decompress_startswith with LZ4 on random text */
compression failed: No buffer space available
/* testing decompress_startswith with LZ4 on HUGE=xxxxxxxxxxxxxxx text */
compression failed: No buffer space available
/* testing LZ4 compression */
/* create source from /build/systemd/src/build/test-compress */
/* test compression */
Buffer size is 589836 bytes, header size 7 bytes.
LZ4 compression finished (22288 -> 9797 bytes, 44.0%)
/* test decompression */
LZ4 decompression finished (9797 -> 22288 bytes, 227.5%)
/* test faulty decompression */
LZ4 decompression finished (9797 -> 22288 bytes, 227.5%)
Decompressed stream longer than 22287 bytes
Compressed 4194304 → 16464
Decompressed → 4194304
Decompressed partial 12/4194304 → 12
Assertion 'r < 0' failed at ../systemd/src/journal/test-compress.c:236, function test_lz4_decompress_partial(). Aborting.
```

**Steps to reproduce the problem**
> Build systemd 239 with 1.8.3 and run test

This appears to have been caused by https://github.com/lz4/lz4/commit/f9725fdc46cc96c49d3cf3905a3b084b27151f33 as using the parent commit the test passes.
I do not know if this is a systemd issue or an lz4 issue so apologies if I am reporting it to the wrong project.

### PR DESCRIPTION
**Pull request by keszybz**
It seems that the function now can decompress partial results into a smaller
buffer. The release news don't say anything interesting, but the test case
that was previously failing now works OK.

Fixes #10259.

### CODE REVIEW

**Comment by poettering** [comment]
Also see #10562

**Comment by keszybz** [comment]
Hmm, I think this actually shows a nasty bug (with the older releases).

**Comment by keszybz** [comment]
Updated. This version is similar at the core to @eworm-de's patch, but uses a dynamic version check instead, and adds a bunch of tests. I thought that the `decompress_startswith()` function was buggy with older lz4s, but it turns out that we were lucky with the buffer sizes and it's not. Nevertheless, I added a bunch of additional testing code.

File: src/journal/compress.c [hunk 1]
```diff
@@ -307,23 +306,37 @@ int decompress_startswith_lz4(const void *src, uint64_t src_size,
...
-        else {
-                /* lz4 always tries to decode full "sequence", so in
-                 * pathological cases might need to decompress the
-                 * full field. */
+        /* One lz4 < 1.8.3, we might get "failure" (r < 0), or "success" where
+         * just a part of the buffer is decompressed. But if we get a smaller
+         * amount of bytes than requested, we don't know whether there isn't enough
+         * data to fill the requested size or whether we just got a partial answer.
+         */
+        if (r < 0 || (unsigned) r < prefix_len + 1) {
```
Comment (poettering) [inline src/journal/compress.c]: not that it matters, but wouldn't be a cast to `size_t` be nicer than `unsigned` here? I mean, `r` is an `int` and you compare it with a `size_t`, hence why pick `unsigned` which is identical to neither type...

Comment (keszybz) [reply src/journal/compress.c]: Updated with that cast changed.

**Pull request merged**